python/lib/core/dmod/core/data_domain_detectors.py

```diff
@@ -0,0 +1,468 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from functools import reduce
+from pathlib import Path
+from typing import TypeVar, Union, Dict, Type, Set, List, Optional
+
+from .common.reader import ReadSeeker
+from .dataset import Dataset
+from .exception import DmodRuntimeError
+from .meta_data import DataDomain, DataFormat, StandardDatasetIndex, DiscreteRestriction
+
+DataItem = TypeVar('DataItem', bound=Union[bytes, ReadSeeker, Path])
+DataCollection = TypeVar('DataCollection', bound=Union[Dataset, Dict[str, Union[bytes, ReadSeeker]], Path])
```

Comment:
That's an interesting idea. We could. Again, it might complicate dealing with local data some. But maybe not, and nothing that would be impossible to deal with.

Reply:
I think it shifts the complexity of dealing with local data to some other code and I think that is a good thing.

Reply:
As I think more about this, this might need to be a TODO for later.

While it probably deserves more discussion, I believe that the rational for removing `Path` would also compel us to remove `bytes`.   And we'll need to update `ItemDataDomainDetector.__init__` to make the `item_name` non-optional.  Also, at that point, we might not really need the _DataItem_ type, and perhaps not the _DataCollection_ type.  Etc.

Good ideas, but things start to snowball a little.  If you feel like they really _need_ to be done before this gets finalized, we can continue discussion.   But otherwise I'd like to table to its own issue.

Reply:
Tracking separately in #583.